### PR TITLE
Support ssr

### PR DESCRIPTION
**Pull request by benjiec**
Adds support for SSR in Edge, specifically supports for CreLox and FLP

### CODE REVIEW

**Review by alecethell — COMMENTED** [review]
Just some minor comments, also, wondering if the plan would be to work through those TODO/FIXMEs now/soon or come back to them later?

File: src/edge/ssr/__init__.py [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+            if event.fragment_id not in old_to_new_fragment_dict:
+                with new_genome.update_fragment_by_fragment_id(
+                    event.fragment_id, new_fragment=True
+                ) as f:
+                    old_to_new_fragment_dict[event.fragment_id] = f
+
+            new_fragment = old_to_new_fragment_dict[event.fragment_id]
+            bp_shift = event.run(new_fragment, self.insert, self.is_insert_circular)
+
+            for future_event in events[1:]:
```
[the remark this answers is not in the document]
Comment (benjiec) [reply src/edge/ssr/__init__.py:596]: hmm, good point =) let me check.

File: src/edge/ssr/__init__.py [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+                    start=loc.start_0based+1
+                ) for loc in self.genomic_locations
+            ]
+        )
+
+
+class IntegrationEvent(Event):
+
+    def get_integrated_aligned_with_site_direction(self, insert, is_insert_circular):
+        insert = insert*3
```
[the remark this answers is not in the document]
Comment (benjiec) [reply src/edge/ssr/__init__.py]: haha, i forgot about flake8, and have a ton of violations. i will fix them.

File: src/edge/models/op.py [hunk 3]
```diff
@@ -8,12 +8,13 @@ class Meta:
 
     # numerical values are stored in database, do not change unless you have a
     # plan for migrating data.
+
     RECOMBINATION = (1, "Homologous Recombination")
     CRISPR_DSB = (2, "CRISPR-Cas9 WT (Double Stranded Break)")
-    PCR_SEQ_VERIFICATION = (3, "PCR Product Sequence Verification")
+    SSR = (4, "Site Specific Recombinase")
```
Comment (davidzhou94) [inline src/edge/models/op.py:14]: For my own curiousity, why did PCR_SEQ_VERIFICATION get removed?

File: src/edge/ssr/crelox.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class Sites(object):
+    loxP     = _c("ATAACTTCGTATA GCATACAT TATACGAAGTTAT")
+    lox66    = _c("ATAACTTCGTATA GCATACAT TATACGAACGGTA")
+    lox71    = _c("TACCGTTCGTATA GCATACAT TATACGAAGTTAT")
+    lox72    = _c("TACCGTTCGTATA GCATACAT TATACGAACGGTA")
+
+    lox5171  = _c("ATAACTTCGTATA GtAcACAT TATACGAAGTTAT")
+    lox2272  = _c("ATAACTTCGTATA GgATACtT TATACGAAGTTAT")
```
Comment (davidzhou94) [inline src/edge/ssr/crelox.py:11]: Why are some of these bases lowercase?

File: src/edge/ssr/__init__.py [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+# sites separated by more than these number of bps will not be considered as
+# part of the same combination
+SITE_SEPARATION_MAX = 50000
+
+
+def rc(s):
+    return str(Seq(s).reverse_complement())
+
+
+def _c(s):
```
Comment (davidzhou94) [inline src/edge/ssr/__init__.py:13]: would be nice to to have a non-single letter name for this like "normalized_sequence" or something

**Review by davidzhou94 — APPROVED** [review]
In general I like the test cases, they helped me understand how this code works. I don't think I really understood how the Reaction objects are able to find sites on the genome, but I think it's not critical for me to understand it now.

Some names could have been more clear. I commented on one example of `_c` but there are others. It's just a nit though.

Nothing looks wrong or in-optimal to me so ship it!

File: src/edge/ssr/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benjiec) [reply src/edge/ssr/__init__.py:642]: works!

File: src/edge/models/op.py (hunk 3 above)
Comment (benjiec) [reply src/edge/models/op.py:14]: i don't think we use that anywhere

File: src/edge/ssr/crelox.py (hunk 4 above)
Comment (benjiec) [reply src/edge/ssr/crelox.py]: those are the mutated sequences from the WT, i am trying to highlight the differences

**Review by alecethell — APPROVED** [review]

**Pull request merged**